### PR TITLE
ngl-diff

### PR DESCRIPTION
**Pull request by cboesch-gpsw**

### CODE REVIEW

File: pynodegl-utils/pynodegl_utils/qml/ngl_widget.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        self._ar = (1, 1)
+
+        self._request_stop = False
+        self._request_scene = None
+        self._livectls_changes = {}
+
+    def createFramebufferObject(self, size):
+        fmt = QOpenGLFramebufferObjectFormat()
+        fmt.setAttachment(QOpenGLFramebufferObject.CombinedDepthStencil)
+        fmt.setSamples(8)
```
Comment (mbouron-gpsw) [inline pynodegl-utils/pynodegl_utils/qml/ngl_widget.py:119]: This should default to 0 (and optionally be configurable though the command line).

File: pynodegl-utils/pynodegl_utils/qml/__init__.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+from PySide6.QtCore import QObject, QUrl
+from PySide6.QtQml import QQmlApplicationEngine, QQmlComponent
+from PySide6.QtQuick import QQuickWindow, QSGRendererInterface
+from PySide6.QtGui import QGuiApplication
+
+
+def create_app_engine(argv, qml_file):
+    # We only support EGL but Qt still uses the legacy GLX by default
+    os.environ["QT_XCB_GL_INTEGRATION"] = "xcb_egl"
+
```
Comment (mbouron-gpsw) [inline pynodegl-utils/pynodegl_utils/qml/__init__.py:35]: ngl-diff will not work as is on macOS as we do not support Opengl 2.x contexts (the default on macOS / QT). You need to enforce the use of OpenGL 4.1. For example you can do something like (which can be extended to support OpenGLES on other platform):
```
surface_format = QtGui.QSurfaceFormat()
surface_format.setRenderableType(QtGui.QSurfaceFormat.OpenGL)
major_version, minor_version = (4, 1)
surface_format.setVersion(major_version, minor_version)
surface_format.setProfile(QtGui.QSurfaceFormat.CoreProfile)
surface_format.setDepthBufferSize(24)
surface_format.setStencilBufferSize(8)
surface_format.setAlphaBufferSize(8)
QtGui.QSurfaceFormat.setDefaultFormat(surface_format)
```

File: pynodegl-utils/pynodegl_utils/qml/ngl_widget.py (hunk 1 above)
Comment (cboesch-gpsw) [reply pynodegl-utils/pynodegl_utils/qml/ngl_widget.py]: Removed for now, thanks, I forgot I had that

File: pynodegl-utils/pynodegl_utils/qml/__init__.py (hunk 2 above)
Comment (cboesch-gpsw) [reply pynodegl-utils/pynodegl_utils/qml/__init__.py:36]: Added, thanks

File: pynodegl-utils/pynodegl_utils/qml/NodeGLWidget.qml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import QMLNodeGL
```
Comment (mbouron-gpsw) [inline pynodegl-utils/pynodegl_utils/qml/NodeGLWidget.qml:1]: Can you add the license header for consistency ?

File: pynodegl-utils/pynodegl_utils/qml/__init__.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+from PySide6.QtGui import QGuiApplication, QSurfaceFormat
+from PySide6.QtQml import QQmlApplicationEngine, QQmlComponent
+from PySide6.QtQuick import QQuickWindow, QSGRendererInterface
+
+
+def create_app_engine(argv, qml_file):
+    # We only support EGL but Qt still uses the legacy GLX by default
+    os.environ["QT_XCB_GL_INTEGRATION"] = "xcb_egl"
+
+    # We do not support Opengl 2.x contexts (the default on macOS Qt), so we
```
Comment (mbouron-gpsw) [inline pynodegl-utils/pynodegl_utils/qml/__init__.py:37]: nit: `2.x` -> `2.1`, `the default on macOS Qt` -> `the default on macOS`, `force OpenGL 4.1` ->  `force OpenGL 4.1 (Core Profile)`

File: pynodegl-utils/pynodegl_utils/qml/NodeGLWidget.qml (hunk 3 above)
Comment (cboesch-gpsw) [reply pynodegl-utils/pynodegl_utils/qml/NodeGLWidget.qml:22]: Done

File: pynodegl-utils/pynodegl_utils/qml/__init__.py (hunk 4 above)
Comment (cboesch-gpsw) [reply pynodegl-utils/pynodegl_utils/qml/__init__.py]: Done

File: pynodegl-utils/pynodegl_utils/qml/ngl_widget.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+            self._request_scene = None
+            assert self._context.set_scene(None) == 0
+            self._t = 0.0
+
+        if self._request_scene:
+            assert self._context.set_scene(self._request_scene) == 0
+            self._request_scene = None
+
+        w, h = self._fbo.width(), self._fbo.height()
+        self._context.gl_wrap_framebuffer(self._fbo.handle())
```
Comment (mbouron-gpsw) [inline pynodegl-utils/pynodegl_utils/qml/ngl_widget.py:151]: Can't this be done only when the fbo has changed ?

Comment (mbouron-gpsw) [reply pynodegl-utils/pynodegl_utils/qml/ngl_widget.py:151]: It looks like it's already done in _NodeGLRender.createFramebufferObject()

File: CHANGELOG.md [hunk 6]
```diff
@@ -11,6 +11,7 @@ Versioning](https://semver.org/spec/v2.0.0.html) for `libnodegl`.
 ### Added
 - Optional `backend_config` field to the `ngl_config` structure (ABI break)
 - External OpenGL context support
+- `ngl-diff` tool to compare videos visually
```
Comment (mbouron-gpsw) [inline CHANGELOG.md:14]: nit: tool to visually compare videos

File: pynodegl-utils/pynodegl_utils/qml/ngl_widget.py (hunk 5 above)
Comment (cboesch-gpsw) [reply pynodegl-utils/pynodegl_utils/qml/ngl_widget.py:151]: It looks like I can't remove any of them for some reason (breaks on resize).

File: CHANGELOG.md (hunk 6 above)
Comment (cboesch-gpsw) [reply CHANGELOG.md]: Done

File: pynodegl-utils/pynodegl_utils/qml/ngl_widget.py (hunk 5 above)
Comment (mbouron-gpsw) [reply pynodegl-utils/pynodegl_utils/qml/ngl_widget.py:158]: I'll investigate that later on.

**Review by mbouron-gpsw — APPROVED** [review]

**Pull request merged**